overlord/snapstate/snapstate.go

```diff
@@ -1396,13 +1394,84 @@ func UpdateMany(ctx context.Context, st *state.State, names []string, revOpts []
-func EnforceSnaps(ctx context.Context, st *state.State, validationSets []string, validationErrorToSolve *snapasserts.ValidationSetsValidationError, userID int) (tasksets []*state.TaskSet, names []string, err error) {
-	// TODO
-	return nil, nil, fmt.Errorf("not implemented")
+// ResolveEnforcementError installs and updates snaps reported by validationErrorToSolve.
+func ResolveEnforcementError(ctx context.Context, st *state.State, vsMap map[string]*asserts.ValidationSet, valErr *snapasserts.ValidationSetsValidationError, userID int) ([]*state.TaskSet, []string, error) {
+	if len(valErr.InvalidSnaps) != 0 {
+		invSnaps := make([]string, 0, len(valErr.InvalidSnaps))
+		for invSnap := range valErr.InvalidSnaps {
+			invSnaps = append(invSnaps, invSnap)
+		}
+		return nil, nil, fmt.Errorf("cannot auto-resolve enforcement constraints that require removing snaps: %s", strutil.Quoted(invSnaps))
+	}
+
+	affected := make([]string, 0, len(valErr.MissingSnaps)+len(valErr.WrongRevisionSnaps))
+	var tasksets []*state.TaskSet
+	// use the same lane for installing and refreshing so everything is reversed
+	lane := st.NewLane()
+
+	collectRevOpts := func(snapToRevToVss map[string]map[snap.Revision][]string) ([]string, []*RevisionOptions) {
+		var names []string
+		var revOpts []*RevisionOptions
+
+		for snapName, revAndVs := range snapToRevToVss {
+			for rev, valsets := range revAndVs {
+				vsKeys := make([]snapasserts.ValidationSetKey, 0, len(valsets))
+				for _, vs := range valsets {
+					vsKey := snapasserts.NewValidationSetKey(valErr.Sets[vs])
+					vsKeys = append(vsKeys, vsKey)
+				}
+
+				revOpts = append(revOpts, &RevisionOptions{Revision: rev, ValidationSets: vsKeys})
+			}
+			names = append(names, snapName)
+		}
+
+		return names, revOpts
+	}
+
+	if len(valErr.WrongRevisionSnaps) > 0 {
+		names, revOpts := collectRevOpts(valErr.WrongRevisionSnaps)
+		flags := &Flags{Transaction: client.TransactionAllSnaps, Lane: lane, NoReRefresh: true}
+
+		updated, tss, err := UpdateMany(ctx, st, names, revOpts, userID, flags)
+		if err != nil {
+			return nil, nil, fmt.Errorf("cannot auto-resolve enforcement constraints: %v", err)
+		}
+
+		tasksets = append(tasksets, tss...)
+		affected = append(affected, updated...)
+	}
+
+	if len(valErr.MissingSnaps) > 0 {
+		names, revOpts := collectRevOpts(valErr.MissingSnaps)
+		flags := &Flags{Transaction: client.TransactionAllSnaps, Lane: lane}
+
+		installed, tss, err := InstallMany(st, names, revOpts, userID, flags)
+		if err != nil {
+			return nil, nil, fmt.Errorf("cannot auto-resolve enforcement constraints: %v", err)
+		}
+
+		tasksets = append(tasksets, tss...)
+		affected = append(affected, installed...)
+	}
+
+	updateTrackingTask := st.NewTask("update-validation-set-tracking", "Update validation sets tracking")
+	updateTrackingTask.Set("userID", userID)
+
+	encodedAsserts := make(map[string][]byte, len(vsMap))
+	for vsStr, vs := range vsMap {
+		encodedAsserts[vsStr] = asserts.Encode(vs)
```

Comment:
didn't we discuss the other day that only encoding the assertion is not enough? I am a bit confused

Reply:
What's missing? If it's information on what sequence point should be pinned, that's encoded in the key.

Reply:
didn't you get error about corresponding account-key or account missing?

Reply:
now I saw how that was solved, given I was just skimming the doc comment on  EnforceValidationSets didn't make me look closer there